### ISSUE TITLE
Built-in dependency that says if the current build is processing any changes

### ISSUE THREAD

**Issue by mbacarella** [ocaml/dune#3159]
Firstly, thanks for making dune.  It roolz!

## Desired Behavior

I'm currently depending on `universe` to cause a build-info file to be re-generated on each dune build.

See more background here: https://discuss.ocaml.org/t/rebuilding-a-build-info-module-every-time-an-exe-is-produced/4954/3

This works, it definitely causes a rebuild if stuff anywhere else in the repo change, which is needed for a proper build-info.  What it also does is rebuild even if nothing changes, which isn't great if you have several hundred .exe files to build.  That is, a make followed by make does a lot of extra work, rebuilding binaries that didn't change.

## Example

Something like, say, `work-to-do` if dune detects any source files have changed instead of `universe` would be grand.

**Comment by diml** [comment]
So it would be a dependency on all the source files in the workspace? Why not, the semantic is clear and it's quite simple to implement.

**Comment by diml** [comment]
Do you want to give it a shot? :)

**Comment by diml** [comment]
Actually, that won't be enough because of external dependencies. If you reinstall a dependency, dune will rebuild things even if the sources haven't changed.

What about a link-time script to generate a module that is executed whenever dune happens to link the executable? Compared to what you propose, it would relink the executable less often. i.e. it wouldn't relink it if something completely unrelated to the executable has changed. But it feels like it would relink it just when it should.

**Comment by mbacarella** [comment]
Hmmm, that sounds like it might be best.  What do you mean by link time
script?  Something with link_deps?

**Comment by diml** [comment]
Not exactly. I'm thinking of the following: you have a module in your executable that has no implementation, and you give dune a command to generate this module at link time. For instance:

```
(executable
 (name foo)
 (link_time_module_gen
  (name build_info)
  (action (run gen_build_info))))
```

and `gen_build_info` would output the contents of `build_info.ml`.

**Comment by mbacarella** [comment]
Seems fine.  Could that directive work on libraries in addition to executables? :smile:

**Comment by diml** [comment]
Sure, that's actually what we do for `dune-build-info` and `findlib.dynload`.

**Comment by mbacarella** [comment]
Sorry, to be clear, is `link_time_module_gen` a feature that already exists or are you proposing something?

**Comment by diml** [comment]
It's an internal feature at the moment. For instance, the  `dune-build-info` library has a module without implementation. The implementation of this module is generated at link time. The stanza for this library looks like this:

```
(library
 (name build_info)
 (public_name dune-build-info)
 (modules_without_implementation build_info_data)
 (special_builtin_support
  (build_info
   (data_module build_info_data)
   (api_version 1))))
```

So the idea would be to generalise this feature so that user can provide their own custom tool to generate the implementation at link time. That would look like this:

```
(library
 (name my_build_info)
 (link_time_code_gen
  (module data)
  (action (run gen-build-info))
  (deps ...)))
```

Then, when an executable depend on this library we would call `gen-build-info` to generate `data.ml` at the same time as linking the executable.

**Comment by mbacarella** [comment]
Right, yeah.  Sweet.  Sounds awesome.

**Comment by diml** [comment]
@mbacarella just to update you on this: @voodoos will be working on this feature.

**Comment by mbacarella** [comment]
Cool, thanks for the update! :D

**Comment by voodoos** [comment]
Hi @mbacarella, I just opened a PR that differs a bit from the proposal sketched above by @jeremiedimino but may work for you: https://github.com/ocaml/dune/pull/3523

It would be great if you could have a look at it, maybe even try it, and tell us whether it solves your issue or not.

**Comment by mbacarella** [comment]
Interesting.  Giving it a shot!
